keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1085 @@
+- Shrinking of volume is not allowed.
+- Few environments require Pod restarts in order to get access to the resized volume (meaning that the underlying CSI driver doesn't support online expansions). This KEP won't support such scenarios.
+- Modifying anything but the `resources.request.storage` field of the `persistentVolumeClaim` (eg, `storageClassName`, `labels`, `annotations`).
+- Resizing a PVC associated with a statefulset directly by modifying the PVC spec will not affect the 
+  Statefulset `.spec.volumeClaimTemplates`. 
+- Resizing ephemeral volumes.
+- Correcting failures resulting from bad input or environmental issues (such as insufficient capacity/quota). We'll make an attempt to do as much validation as we can in the API server, but cases of errors/failures past admission will have to be handled by the user. This KEP will discuss the different failures that could happen past admission.
+
+  
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Before we proceed to the proposal it is imperative to layout the various components involved
+in creating and managing the persistent volumes associated with `StatefulSet`. 
+
+- StatefulSet controller creates the PVCs from `volumeClaimTemplates` and associates them with the pods
+spawned from it. Reconciliation logic runs for every change made to the StatefulSet
+`Spec` and `Status` reflects the progress towards making the real world state equal to the
+`Spec`.
+- The validation logic in the API server is responsible for restricting the fields
+which can be edited in a StatefulSet specification.
+  - Part of the validation that the API server performs for objects (eg StatefulSet, PVC) happens in the object API server code and part happens in an admission plugin.
+- The RBAC specification in the controller policy is responsible for adding the required 
+RBAC permission for StatefulSet.
+- The StatefulSet controller also supports revision control.
+
+The following changes are proposed to achieve the goal:
+1. Make the `volumeClaimTemplates` storage size (`.spec.resources.request.storage`) editable by modification to the api server
+   validation. We should allow the user to both increase or decrease the storage size. Allowing the user to decrease the storage size will give them the ability to address failed expansions, as described in [KEP 1790](../1790-recover-resize-failure/README.md).
+2. Add reconciliation of the associated PVC's size and the individual `volumeClaimTemplates` size into the 
+   StatefulSet controller reconciliation.
+3. Add PVC `patch` privilege to the StatefulSet controller's clusterrole RBAC.
```

Comment:
Can we also mention here that, we will be changing API for `StatefulSet.Status` ?

Reply:
will do